File: stdlib/lua/hull/middleware/tests/test_ratelimit.lua
-- test_ratelimit.lua — Tests for hull.middleware.ratelimit
--
-- Tests pure-function helpers (no runtime globals needed).
-- Run via: the C test harness (test_lua_runtime.c) loads and executes this.

local ratelimit = require('hull.middleware.ratelimit')

local pass = 0
local fail = 0

local function test(name, fn)
    local ok, err = pcall(fn)
    if ok then
        pass = pass + 1
    else
        fail = fail + 1
        print("FAIL: " .. name .. ": " .. tostring(err))
    end
end

local function assert_eq(a, b, msg)
    if a ~= b then
        error((msg or "") .. " expected " .. tostring(b) .. ", got " .. tostring(a))
    end
end

-- ── check ────────────────────────────────────────────────────────────

test("check allows first request", function()
    local buckets = {}
    local r = ratelimit.check(buckets, "ip1", 5, 60, 1000)
    assert_eq(r.allowed, true)
    assert_eq(r.remaining, 4)
end)

test("check blocks after limit", function()
    local buckets = {}
    for i = 1, 5 do
        ratelimit.check(buckets, "ip2", 5, 60, 1000)
    end
    local r = ratelimit.check(buckets, "ip2", 5, 60, 1000)
    assert_eq(r.allowed, false)
    assert_eq(r.remaining, 0)
end)

test("check resets after window", function()
    local buckets = {}
    for i = 1, 5 do
        ratelimit.check(buckets, "ip3", 5, 60, 1000)
    end
    local r = ratelimit.check(buckets, "ip3", 5, 60, 1061)
    assert_eq(r.allowed, true)
    assert_eq(r.remaining, 4)
end)

test("check returns remaining count", function()
    local buckets = {}
    ratelimit.check(buckets, "ip4", 10, 60, 1000)
    ratelimit.check(buckets, "ip4", 10, 60, 1000)
    local r = ratelimit.check(buckets, "ip4", 10, 60, 1000)
    assert_eq(r.remaining, 7)
end)

test("check returns reset timestamp", function()
    local buckets = {}
    local r = ratelimit.check(buckets, "ip5", 5, 60, 1000)
    assert_eq(r.reset, 1060)
end)

-- ── middleware factory ───────────────────────────────────────────────

test("middleware returns a function", function()
    local mw = ratelimit.middleware({ limit = 10, window = 60 })
    assert(type(mw) == "function", "expected function")
end)

return {pass = pass, fail = fail}

File: stdlib/js/hull/middleware/tests/test_ratelimit.js
// test_ratelimit.js — Tests for hull:middleware:ratelimit
//
// Tests pure-function helpers (no runtime globals needed).

import { ratelimit } from "hull:middleware:ratelimit";

let pass = 0;
let fail = 0;

function test(name, fn) {
    try {
        fn();
        pass++;
    } catch (e) {
        fail++;
        print("FAIL: " + name + ": " + e.message);
    }
}

function assertEq(a, b, msg) {
    if (a !== b)
        throw new Error((msg || "") + " expected " + b + ", got " + a);
}

// ── check ────────────────────────────────────────────────────────────

test("check allows first request", () => {
    const buckets = {};
    const r = ratelimit.check(buckets, "ip1", 5, 60, 1000);
    assertEq(r.allowed, true);
    assertEq(r.remaining, 4);
});

test("check blocks after limit", () => {
    const buckets = {};
    for (let i = 0; i < 5; i++)
        ratelimit.check(buckets, "ip2", 5, 60, 1000);
    const r = ratelimit.check(buckets, "ip2", 5, 60, 1000);
    assertEq(r.allowed, false);
    assertEq(r.remaining, 0);
});

test("check resets after window", () => {
    const buckets = {};
    for (let i = 0; i < 5; i++)
        ratelimit.check(buckets, "ip3", 5, 60, 1000);
    const r = ratelimit.check(buckets, "ip3", 5, 60, 1061);
    assertEq(r.allowed, true);
    assertEq(r.remaining, 4);
});

test("check returns remaining count", () => {
    const buckets = {};
    ratelimit.check(buckets, "ip4", 10, 60, 1000);
    ratelimit.check(buckets, "ip4", 10, 60, 1000);
    const r = ratelimit.check(buckets, "ip4", 10, 60, 1000);
    assertEq(r.remaining, 7);
});

test("check returns reset timestamp", () => {
    const buckets = {};
    const r = ratelimit.check(buckets, "ip5", 5, 60, 1000);
    assertEq(r.reset, 1060);
});

// ── middleware factory ───────────────────────────────────────────────

test("middleware returns a function", () => {
    const mw = ratelimit.middleware({ limit: 10, window: 60 });
    assertEq(typeof mw, "function");
});

export default { pass, fail };
